### ISSUE TITLE
Export Python functions to ONNX as a single op

### ISSUE THREAD

**Issue by gyenesvi** [onnx/onnx#3286]
# Ask a Question

### Question
I have seen some (experimental?) support in ONNX proto for functions (sequence of ops). Is it actually supported by ONNX latest version, or was it just experimental? I could not find any documentation or examples on it. If supported, how can I export a PyTorch model with such a function in it?

### Further information
- Relevant Area: model usage, best practices
- Is this issue related to a specific model?  
No

### Notes
For example, when exporting a ShuffleNet, it would be good to have the shuffle op as a single op/function so that it is easier on the importer side to understand which ops form a higher level compound. True that for small functions like the shuffle op, it could be merged on the importer side, but such practice becomes prohibitive for larger functions.

**Comment by jcwchen** [comment]
Hi @gyenesvi,
ONNX supports introducing a function op which consists of existing ONNX operators. As far as I know, there is no proper document for it now so you can check [softmax](https://github.com/onnx/onnx/blob/4f7e1f6e8bac5e5b28cd8ce7553f822d332e03bc/onnx/defs/math/defs.cc#L1196) as example. Here is the related [PR](https://github.com/onnx/onnx/pull/2573). Typically, if such a function op has been proposed in ONNX, then all of the converters will try to support this conversion from the original framework. Thank you.

**Comment by gyenesvi** [comment]
Hi @jcwchen,
Thanks for the references. So does this mean that those functions can only be used when they are specified as such in ONNX? Or is it possible to create functions on-demand during export? Again, the question is not whether a specific exporter can do that, rather whether ONNX is capable of representing that? What I'd like to be able to do is let a user of an exporter specify an arbitrary sub-graph and put it into a function and save it like that into ONNX. This sub-graph could be quite large and task specific sub-computation so it would be inappropriate to specify it as a function by the ONNX opset itself, that's why I say it's on-demand.

**Comment by jcwchen** [comment]
Theoretically, ONNX function op can consist of undefined operators because it does not reply on onnx's checker or shape_inference. https://github.com/onnx/onnx/blob/490895703c82f68e723a8b308369cac4ed0abc0c/onnx/defs/function.cc#L112 So I would say this question only depends on whether targeted converter supports the conversion of subgraph to unofficial ONNX operators without touching `onnx.shape_inference`. Thanks.

**Comment by gyenesvi** [comment]
@jcwchen I am not asking about undefined or custom operators in a function. I'd like to export an arbitrary sub-graph made of official ONNX operators into a function. So with such operators, can ONNX describe an arbitrary sub-graph as a function, or do functions need to be a pre-defined in ONNX? Thanks

**Comment by jcwchen** [comment]
Hi @gyenesvi,
All single operators and function operators need to be pre-defined under `onnx/defs/`. They cannot be defined on-the-fly. Thanks.

**Comment by gyenesvi** [comment]
Thanks, it's clear now.
